x-pack/filebeat/input/gcs/job.go

```diff
@@ -109,12 +109,14 @@ func (j *job) do(ctx context.Context, id string) {
 			Fields:    fields,
 		}
 		event.SetID(objectID(j.hash, 0))
-		j.state.save(j.object.Name, j.object.Updated)
-		// locks while data is being published to avoid concurrent map read/writes
+		// locks while data is being saved and published to avoid concurrent map read/writes
 		j.mu.Lock()
-		if err := j.publisher.Publish(event, j.state.checkpoint()); err != nil {
+		j.state.save(j.object.Name, j.object.Updated)
+		cp, done := j.state.checkpointTxn()
```

Comment:
I had more of a think about this. If we have 
```
// save, saves/updates the current state for cursor checkpoint
func (s *state) save(name string, lastModifiedOn time.Time) {
	_, done := s.saveForTx(name, lastModifiedOn)
	done()
}

// saveForTx updates and returns the current state checkpoint, locks the state
// and returns an unlock function, done. The caller must call done when
// cp is no longer needed in a locked state. done may not be called
// more than once.
func (s *state) saveForTx(name string, lastModifiedOn time.Time) (cp *Checkpoint, done func()) {
	s.mu.Lock()
	delete(s.cp.LastProcessedOffset, name)
	delete(s.cp.IsRootArray, name)
	if _, ok := s.cp.FailedJobs[name]; !ok {
		if len(s.cp.ObjectName) == 0 {
			s.cp.ObjectName = name
		} else if strings.ToLower(name) > strings.ToLower(s.cp.ObjectName) {
			s.cp.ObjectName = name
		}

		if s.cp.LatestEntryTime.IsZero() {
			s.cp.LatestEntryTime = lastModifiedOn
		} else if lastModifiedOn.After(s.cp.LatestEntryTime) {
			s.cp.LatestEntryTime = lastModifiedOn
		}
	} else {
		// clear entry if this is a failed job
		delete(s.cp.FailedJobs, name)
	}
	return s.cp, func() { s.mu.Unlock() }
}
```
then j.state.save/j.state.checkpointTxn merge into a single call/lock with j.state.saveForTxn and j.mu should no longer be needed.

Reply:
Similar change for `savePartial` for below. It looks like `save` (and `savePartial`) is always used before a publish, so really we don't need the self-re-unlocking variant, though I'd use the new name for clarity.